src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -71,92 +71,148 @@ private class ConfigurationCacheItem
         }
 
         public ValueTask<bool> IsEnabledAsync<TContext>(string feature, TContext appContext, CancellationToken cancellationToken)
         {
-            return IsEnabledWithVariantsAsync(feature, appContext, useAppContext: true, cancellationToken);
+            return IsEnabledEvaluation(feature, appContext, useAppContext: true, cancellationToken);
         }
 
-        private async ValueTask<bool> IsEnabledWithVariantsAsync<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
+        private async ValueTask<bool> IsEnabledEvaluation<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
         {
-            bool isFeatureEnabled = false;
+            EvaluationEvent evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
 
-            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);
+            await EvaluateFeature(evaluationEvent, appContext, useAppContext, cancellationToken);
 
-            VariantDefinition variantDefinition = null;
+            return evaluationEvent.IsEnabled;
+        }
 
-            if (featureDefinition != null)
+        public ValueTask<Variant> GetVariantAsync(string feature, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(feature))
             {
-                isFeatureEnabled = await IsEnabledAsync(featureDefinition, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            return GetVariantEvaluation(feature, context: null, useContext: false, cancellationToken);
+        }
+
+        public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext context, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(feature))
+            {
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            return GetVariantEvaluation(feature, context, useContext: true, cancellationToken);
+        }
 
-                if (featureDefinition.Variants != null && featureDefinition.Variants.Any() && featureDefinition.Allocation != null)
+        private async ValueTask<Variant> GetVariantEvaluation(string feature, TargetingContext context, bool useContext, CancellationToken cancellationToken)
+        {
+            EvaluationEvent evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
+
+            await EvaluateFeature(evaluationEvent, context, useContext, cancellationToken);
```

Comment:
It looks like `EvaluateFeature` returns an `EvaluationEvent` but the return value isn't used anywhere it's called.

Reply:
Yeah that's accurate. I wanted to indicate that it's the resulting object (all of the results are on the object), but it's the same object that's passed in, so these could all be void.

Reply:
Since it is the same object that's passed in, we should use `ref` parameter. Unfortunately, this is an async method, we cannot use `ref` parameters.
I suggest:
```C#
private async Task<EvaluationEvent> EvaluateFeature<TContext>(string feature, TContext context, bool useContext, CancellationToken cancellationToken)
{
   EvaluationEvent evaluationEvent = new EvaluationEvent
   {
       FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
   };

   ...
}
```